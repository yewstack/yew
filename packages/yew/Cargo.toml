[package]
name = "yew"
version = "0.17.4"
edition = "2018"
authors = [
  "Denis Kolodin <deniskolodin@gmail.com>",
  "Justin Starry <justin@yew.rs>",
]
repository = "https://github.com/yewstack/yew"
homepage = "https://github.com/yewstack/yew"
documentation = "https://docs.rs/yew/"
license = "MIT/Apache-2.0"
readme = "../README.md"
keywords = ["web", "webasm", "javascript"]
categories = ["gui", "wasm", "web-programming"]
description = "A framework for making client-side single-page apps"

[dependencies]
anyhow = "1"
anymap = "0.12"
console_error_panic_hook = "0.1"
gloo = "0.2.1"
http = "0.2"
indexmap = { version = "1.5", features = ["std"] }
js-sys = "0.3"
log = "0.4"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
slab = "0.4"
thiserror = "1"
wasm-bindgen = "0.2.60"
wasm-bindgen-futures = "0.4"
yew-macro = { version = "0.17.0", path = "../yew-macro" }

# optional encodings
bincode = { version = "1", optional = true }
rmp-serde = { version = "0.15.0", optional = true }
serde_cbor = { version = "0.11.1", optional = true }
serde_yaml = { version = "0.8.4", optional = true }
toml = { version = "0.5", optional = true }

[dependencies.web-sys]
version = "0.3"
features = [
  "AbortController",
  "AbortSignal",
  "AnimationEvent",
  "BinaryType",
  "Blob",
  "BlobPropertyBag",
  "console",
  "DedicatedWorkerGlobalScope",
  "Document",
  "DomTokenList",
  "DragEvent",
  "Element",
  "ErrorEvent",
  "Event",
  "EventTarget",
  "File",
  "FileList",
  "FileReader",
  "FocusEvent",
  "Headers",
  "HtmlElement",
  "HtmlButtonElement",
  "HtmlInputElement",
  "HtmlSelectElement",
  "HtmlTextAreaElement",
  "InputEvent",
  "KeyboardEvent",
  "Location",
  "MessageEvent",
  "MouseEvent",
  "Node",
  "ObserverCallback",
  "PointerEvent",
  "ProgressEvent",
  "ReferrerPolicy",
  "Request",
  "RequestCache",
  "RequestCredentials",
  "RequestInit",
  "RequestMode",
  "RequestRedirect",
  "Response",
  "Storage",
  "Text",
  "TouchEvent",
  "TransitionEvent",
  "UiEvent",
  "Url",
  "WebSocket",
  "WheelEvent",
  "Window",
  "Worker",
  "WorkerGlobalScope",
  "WorkerOptions",
]

[dev-dependencies]
base64 = "0.13.0"
bincode = "1"
easybench-wasm = "0.2.1"
rmp-serde = "0.15.0"
rustversion = "1.0"
serde_derive = "1"
ssri = "6.0.0"
trybuild = "1.0"
wasm-bindgen-test = "0.3.4"

[features]
default = ["agent"]
doc_test = []
wasm_test = []
wasm_bench = []
agent = ["bincode"]
yaml = ["serde_yaml"]
msgpack = ["rmp-serde"]
cbor = ["serde_cbor"]

[package.metadata.docs.rs]
features = ["yaml", "cbor", "toml", "msgpack", "doc_test"]
