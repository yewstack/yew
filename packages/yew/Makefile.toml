[env]
CARGO_MAKE_CLIPPY_ARGS = "--features cbor,msgpack,toml,yaml -- --deny=warnings"

[tasks.test]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "@@split(YEW_TEST_FLAGS, )",
    "--",
    "--features",
    "${YEW_TEST_FEATURES}",
]

[tasks.doc-test]
clear = true
run_task = { name = ["doc-test-normal"], fork = true }

[tasks.doc-test-normal]
command = "cargo"
args = [
    "test",
    "--doc",
    "--features",
    "doc_test,wasm_test,yaml,msgpack,cbor,toml",
]

[tasks.bench]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "--release",
    "--firefox",
    "--headless",
    "--",
    "--features",
    "wasm_bench",
    "bench",
]
