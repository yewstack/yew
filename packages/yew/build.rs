use std::env;

pub fn main() {
    let using_web_sys = cfg!(feature = "web_sys");
    let using_std_web = cfg!(feature = "std_web");
    if using_web_sys && using_std_web {
        panic!("Yew does not allow the `web_sys` and `std_web` cargo features to be used simultaneously");
    } else if !using_web_sys && !using_std_web {
        panic!("Yew requires selecting either the `web_sys` or `std_web` cargo feature");
    }

    let using_cargo_web = env::var("COMPILING_UNDER_CARGO_WEB").is_ok();
    if using_cargo_web && using_web_sys {
        panic!("cargo-web is not compatible with web-sys");
    }
}
