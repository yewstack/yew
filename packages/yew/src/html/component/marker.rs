//! Primitive Components & Properties Types

use crate::html::{BaseComponent, ChildrenProps, Html};
use crate::{function_component, html};

/// A Component to represent a component that does not exist in current implementation.
///
/// During Hydration, Yew expected the Virtual DOM hierarchy to match the the layout used in
/// server-side rendering. However, sometimes it is possible / reasonable to omit certain components
/// from one side of the implementation. This component is used to represent a component as if a
/// component "existed" in the place it is defined.
///
/// # Warning
///
/// The Real DOM hierarchy must also match the server-side rendered artifact. This component is
/// only usable when the original component does not introduce any additional elements. (e.g.:
/// Context Providers)
///
/// A generic parameter is provided to help identify the component to be substituted.
/// The type of the generic parameter is not required to be the same component that was in the other
/// implementation. However, this behaviour may change in the future if more debug assertions were
/// to be introduced. It is recommended that the generic parameter represents the component in the
/// other implementation.
///
/// # Example
///
/// ```
/// use yew::prelude::*;
/// # use yew::html::ChildrenProps;
/// #
/// # #[function_component]
/// # fn Comp(props: &ChildrenProps) -> Html {
/// #     Html::default()
/// # }
/// #
/// # #[function_component]
/// # fn Provider(props: &ChildrenProps) -> Html {
/// #     let children = props.children.clone();
/// #
/// #     html! { <>{children}</> }
/// # }
/// # type Provider1 = Provider;
/// # type Provider2 = Provider;
/// # type Provider3 = Provider;
/// # type Provider4 = Provider;
///
/// #[function_component]
/// fn ServerApp() -> Html {
///     // The Server Side Rendering Application has 3 Providers.
///     html! {
///         <Provider1>
///             <Provider2>
///                 <Provider3>
///                     <Comp />
///                 </Provider3>
///             </Provider2>
///         </Provider1>
///     }
/// }
///
/// #[function_component]
/// fn App() -> Html {
///     // The Client Side Rendering Application has 4 Providers.
///     html! {
///         <Provider1>
///             <Provider2>
///                 <Provider3>
///
///                     // This provider does not exist on the server-side
///                     // Hydration will fail due to Virtual DOM layout mismatch.
///                     <Provider4>
///                         <Comp />
///                     </Provider4>
///
///                 </Provider3>
///             </Provider2>
///         </Provider1>
///     }
/// }
/// ```
///
/// To mitigate this, we can use a `PhantomComponent`:
///
/// ```
/// use yew::prelude::*;
/// # use yew::html::{PhantomComponent, ChildrenProps};
/// #
/// # #[function_component]
/// # fn Comp(props: &ChildrenProps) -> Html {
/// #     Html::default()
/// # }
/// #
/// # #[function_component]
/// # fn Provider(props: &ChildrenProps) -> Html {
/// #     let children = props.children.clone();
/// #
/// #     html! { <>{children}</> }
/// # }
/// # type Provider1 = Provider;
/// # type Provider2 = Provider;
/// # type Provider3 = Provider;
/// # type Provider4 = Provider;
///
/// #[function_component]
/// fn ServerApp() -> Html {
///     html! {
///         <Provider1>
///             <Provider2>
///                 <Provider3>
///                     // We add a PhantomComponent for Provider4,
///                     // it acts if a Provider4 component presents in this position.
///                     <PhantomComponent<Provider4>>
///                         <Comp />
///                     </PhantomComponent<Provider4>>
///                 </Provider3>
///             </Provider2>
///         </Provider1>
///     }
/// }
///
/// #[function_component]
/// fn App() -> Html {
///     html! {
///         <Provider1>
///             <Provider2>
///                 <Provider3>
///
///                     // Hydration will succeed as the PhantomComponent in the server-side
///                     // implementation will represent a Provider4 component in this position.
///                     <Provider4>
///                         <Comp />
///                     </Provider4>
///
///                 </Provider3>
///             </Provider2>
///         </Provider1>
///     }
/// }
/// ```
#[function_component]
pub fn PhantomComponent<T>(props: &ChildrenProps) -> Html
where
    T: BaseComponent,
{
    html! { <>{props.children.clone()}</> }
}
