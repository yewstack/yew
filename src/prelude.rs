//! The Yew Prelude
//!
//! The purpose of this module is to alleviate imports of many common types:
//!
//! ```
//! # #![allow(unused_imports)]
//! use yew::prelude::*;
//! ```
extern crate stdweb;

pub use html::{
    Component,
    ComponentLink,
    Href,
    Html,
    ChangeData,
    InputData,
    Renderable,
    ShouldRender,
};

pub use app::App;

pub use callback::Callback;

pub use stdweb::web::event::{
    BlurEvent,
    ClickEvent,
    DoubleClickEvent,
    IKeyboardEvent,
    IMouseEvent,
    KeyDownEvent,
    KeyPressEvent,
    KeyUpEvent,
    MouseDownEvent,
    MouseMoveEvent,
    MouseOutEvent,
    MouseOverEvent,
    MouseUpEvent,
};

pub use agent::{
    Ambit,
    Bridge,
    Worker,
};

/// Prelude module for creating worker.
pub mod worker {
    pub use agent::{
        Addr,
        Agent,
        AgentLink,
        HandlerId,
        Transferable,
        Worker,
    };
}
