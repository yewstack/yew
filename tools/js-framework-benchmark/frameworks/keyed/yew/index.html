<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <title>Yew</title>
    <link href="/css/currentStyle.css" rel="stylesheet"/>
    <base href="bundled-dist/"></base>
  </head>
  <body>
    <div id='main'></div>
    <script type="module">
      import init from './js-framework-benchmark-yew.js';
      init('./js-framework-benchmark-yew_bg.wasm');
    </script>
  </body>
</html>
