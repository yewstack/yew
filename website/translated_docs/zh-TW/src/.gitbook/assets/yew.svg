<svg width="75" height="82" viewBox="0 0 75 82" fill="none" xmlns="http://www.w3.org/2000/svg">
<circle cx="38" cy="40" r="25" fill="#FFEFB8"/>
<path d="M38.2373 41.0339L14 14" stroke="#444444" stroke-width="6" stroke-linecap="round"/>
<path d="M38.2373 41.0339L62.4746 14" stroke="#444444" stroke-width="6" stroke-linecap="round"/>
<path d="M38.2373 41.0339L38.2373 69" stroke="#444444" stroke-width="6" stroke-linecap="round"/>
<circle cx="38" cy="41" r="7" fill="#FFD707" stroke="#444444" stroke-width="4"/>
</svg>
