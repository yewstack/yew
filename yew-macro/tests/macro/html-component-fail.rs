#![recursion_limit = "128"]

use std::marker::PhantomData;
use yew::prelude::*;

#[derive(Clone, Properties, PartialEq)]
pub struct ChildProperties {
    #[prop_or_default]
    pub string: String,
    pub int: i32,
}

pub struct Child;
impl Component for Child {
    type Message = ();
    type Properties = ChildProperties;

    fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self {
        unimplemented!()
    }
    fn update(&mut self, _: Self::Message) -> ShouldRender {
        unimplemented!()
    }
    fn change(&mut self, _: Self::Properties) -> ShouldRender {
        unimplemented!()
    }
    fn view(&self) -> Html {
        unimplemented!()
    }
}

#[derive(Clone, Properties)]
pub struct ChildContainerProperties {
    pub children: ChildrenWithProps<Child>,
}

pub struct ChildContainer;
impl Component for ChildContainer {
    type Message = ();
    type Properties = ChildContainerProperties;

    fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self {
        unimplemented!()
    }
    fn update(&mut self, _: Self::Message) -> ShouldRender {
        unimplemented!()
    }
    fn change(&mut self, _: Self::Properties) -> ShouldRender {
        unimplemented!()
    }
    fn view(&self) -> Html {
        unimplemented!()
    }
}

pub struct Generic<G> {
    marker: PhantomData<G>,
}

impl Component for Generic<String> {
    type Message = ();
    type Properties = ();

    fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self {
        unimplemented!()
    }
    fn update(&mut self, _: Self::Message) -> ShouldRender {
        unimplemented!()
    }
    fn change(&mut self, _: Self::Properties) -> ShouldRender {
        unimplemented!()
    }
    fn view(&self) -> Html {
        unimplemented!()
    }
}

fn compile_fail() {
    html! { <Child> };
    html! { <Child:: /> };
    html! { <Child with /> };
    html! { <Child props /> };
    html! { <Child with props > };
    html! { <Child with props ref=() ref=() /> };
    html! { <Child with props ref=() ref=() value=1 /> };
    html! { <Child with props ref=() value=1 ref=() /> };
    html! { <Child with props value=1 ref=()  ref=() /> };
    html! { <Child value=1 with props  ref=()  ref=() /> };
    html! { <Child value=1 ref=() with props ref=() /> };
    html! { <Child ref=() ref=() value=1  with props  /> };
    html! { <Child with blah /> };
    html! { <Child with props () /> };
    html! { <Child value=1 with props /> };
    html! { <Child with props value=1 /> };
    html! { <Child type=0 /> };
    html! { <Child invalid-prop-name=0 /> };
    html! { <Child unknown="unknown" /> };
    html! { <Child string= /> };
    html! { <Child int=1 string={} /> };
    html! { <Child int=1 string=3 /> };
    html! { <Child int=1 string={3} /> };
    html! { <Child int=1 ref=() /> };
    html! { <Child int=1 ref=() ref=() /> };
    html! { <Child int=0u32 /> };
    html! { <Child string="abc" /> };
    html! { </Child> };
    html! { <Child><Child></Child> };
    html! { <Child></Child><Child></Child> };
    html! { <Child>{ "Not allowed" }</Child> };

    html! { <ChildContainer /> };
    html! { <ChildContainer></ChildContainer> };
    html! { <ChildContainer>{ "Not allowed" }</ChildContainer> };
    html! { <ChildContainer><></></ChildContainer> };
    html! { <ChildContainer><other /></ChildContainer> };

    html! { <Generic<String>></Generic> };
    html! { <Generic<String>></Generic<Vec<String>>> };

    html_nested! {
        <span>{ 1 }</span>
        <span>{ 2 }</span>
    };

    html! { <TestComponent value?="not_supported" /> };
}

fn main() {}
