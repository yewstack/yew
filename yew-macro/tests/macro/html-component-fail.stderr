error: this opening tag has no corresponding closing tag
  --> $DIR/html-component-fail.rs:79:13
   |
79 |     html! { <Child> };
   |             ^^^^^^^

error: expected identifier
  --> $DIR/html-component-fail.rs:80:22
   |
80 |     html! { <Child:: /> };
   |                      ^

error: `with` must be followed by an identifier
  --> $DIR/html-component-fail.rs:81:20
   |
81 |     html! { <Child with /> };
   |                    ^^^^

error: this prop doesn't have a value
  --> $DIR/html-component-fail.rs:82:20
   |
82 |     html! { <Child props /> };
   |                    ^^^^^

error: this opening tag has no corresponding closing tag
  --> $DIR/html-component-fail.rs:83:13
   |
83 |     html! { <Child with props > };
   |             ^^^^^^^^^^^^^^^^^^^

error: too many refs set
  --> $DIR/html-component-fail.rs:84:38
   |
84 |     html! { <Child with props ref=() ref=() /> };
   |                                      ^^^

error: too many refs set
  --> $DIR/html-component-fail.rs:85:38
   |
85 |     html! { <Child with props ref=() ref=() value=1 /> };
   |                                      ^^^

error: Using the `with props` syntax in combination with named props is not allowed (note: this does not apply to the `ref` prop).
  --> $DIR/html-component-fail.rs:86:38
   |
86 |     html! { <Child with props ref=() value=1 ref=() /> };
   |                                      ^^^^^

error: Using the `with props` syntax in combination with named props is not allowed (note: this does not apply to the `ref` prop).
  --> $DIR/html-component-fail.rs:87:31
   |
87 |     html! { <Child with props value=1 ref=()  ref=() /> };
   |                               ^^^^^

error: Using the `with props` syntax in combination with named props is not allowed (note: this does not apply to the `ref` prop).
  --> $DIR/html-component-fail.rs:88:28
   |
88 |     html! { <Child value=1 with props  ref=()  ref=() /> };
   |                            ^^^^

error: Using the `with props` syntax in combination with named props is not allowed (note: this does not apply to the `ref` prop).
  --> $DIR/html-component-fail.rs:89:35
   |
89 |     html! { <Child value=1 ref=() with props ref=() /> };
   |                                   ^^^^

error: too many refs set
  --> $DIR/html-component-fail.rs:90:27
   |
90 |     html! { <Child ref=() ref=() value=1  with props  /> };
   |                           ^^^

error: unexpected token
  --> $DIR/html-component-fail.rs:92:31
   |
92 |     html! { <Child with props () /> };
   |                               ^^

error: Using the `with props` syntax in combination with named props is not allowed (note: this does not apply to the `ref` prop).
  --> $DIR/html-component-fail.rs:93:28
   |
93 |     html! { <Child value=1 with props /> };
   |                            ^^^^

error: Using the `with props` syntax in combination with named props is not allowed (note: this does not apply to the `ref` prop).
  --> $DIR/html-component-fail.rs:94:31
   |
94 |     html! { <Child with props value=1 /> };
   |                               ^^^^^

error: expected identifier
  --> $DIR/html-component-fail.rs:95:20
   |
95 |     html! { <Child type=0 /> };
   |                    ^^^^

error: expected identifier
  --> $DIR/html-component-fail.rs:96:20
   |
96 |     html! { <Child invalid-prop-name=0 /> };
   |                    ^^^^^^^^^^^^^^^^^

error: expected an expression following this equals sign
  --> $DIR/html-component-fail.rs:98:26
   |
98 |     html! { <Child string= /> };
   |                          ^

error: too many refs set
   --> $DIR/html-component-fail.rs:103:33
    |
103 |     html! { <Child int=1 ref=() ref=() /> };
    |                                 ^^^

error: this closing tag has no corresponding opening tag
   --> $DIR/html-component-fail.rs:106:13
    |
106 |     html! { </Child> };
    |             ^^^^^^^^

error: this opening tag has no corresponding closing tag
   --> $DIR/html-component-fail.rs:107:13
    |
107 |     html! { <Child><Child></Child> };
    |             ^^^^^^^

error: only one root html element is allowed (hint: you can wrap multiple html elements in a fragment `<></>`)
   --> $DIR/html-component-fail.rs:108:28
    |
108 |     html! { <Child></Child><Child></Child> };
    |                            ^^^^^^^^^^^^^^^

error: this closing tag has no corresponding opening tag
   --> $DIR/html-component-fail.rs:117:30
    |
117 |     html! { <Generic<String>></Generic> };
    |                              ^^^^^^^^^^

error: this closing tag has no corresponding opening tag
   --> $DIR/html-component-fail.rs:118:30
    |
118 |     html! { <Generic<String>></Generic<Vec<String>>> };
    |                              ^^^^^^^^^^^^^^^^^^^^^^^

error: only one root html element is allowed (hint: you can wrap multiple html elements in a fragment `<></>`)
   --> $DIR/html-component-fail.rs:122:9
    |
122 |         <span>{ 2 }</span>
    |         ^^^^^^^^^^^^^^^^^^

error: optional attributes are only supported on HTML tags. Yew components can use `Option<T>` properties to accomplish the same thing.
   --> $DIR/html-component-fail.rs:125:28
    |
125 |     html! { <TestComponent value?="not_supported" /> };
    |                            ^^^^^

error[E0425]: cannot find value `blah` in this scope
  --> $DIR/html-component-fail.rs:91:25
   |
91 |     html! { <Child with blah /> };
   |                         ^^^^ not found in this scope

error[E0609]: no field `unknown` on type `ChildProperties`
  --> $DIR/html-component-fail.rs:97:20
   |
97 |     html! { <Child unknown="unknown" /> };
   |                    ^^^^^^^ unknown field
   |
   = note: available fields are: `string`, `int`

error[E0599]: no method named `unknown` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
  --> $DIR/html-component-fail.rs:97:20
   |
6  | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `unknown` not found for this
...
97 |     html! { <Child unknown="unknown" /> };
   |                    ^^^^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VComp: yew::virtual_dom::Transformer<(), std::string::String>` is not satisfied
  --> $DIR/html-component-fail.rs:99:33
   |
99 |     html! { <Child int=1 string={} /> };
   |                                 ^^ the trait `yew::virtual_dom::Transformer<(), std::string::String>` is not implemented for `yew::virtual_dom::vcomp::VComp`
   |
   = help: the following implementations were found:
             <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, T>>
             <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, std::option::Option<T>>>
             <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::option::Option<std::string::String>>>
             <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::string::String>>
           and 3 others
   = note: required by `yew::virtual_dom::Transformer::transform`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VComp: yew::virtual_dom::Transformer<{integer}, std::string::String>` is not satisfied
   --> $DIR/html-component-fail.rs:100:33
    |
100 |     html! { <Child int=1 string=3 /> };
    |                                 ^ the trait `yew::virtual_dom::Transformer<{integer}, std::string::String>` is not implemented for `yew::virtual_dom::vcomp::VComp`
    |
    = help: the following implementations were found:
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, T>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, std::option::Option<T>>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::option::Option<std::string::String>>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::string::String>>
            and 3 others
    = note: required by `yew::virtual_dom::Transformer::transform`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VComp: yew::virtual_dom::Transformer<{integer}, std::string::String>` is not satisfied
   --> $DIR/html-component-fail.rs:101:33
    |
101 |     html! { <Child int=1 string={3} /> };
    |                                 ^^^ the trait `yew::virtual_dom::Transformer<{integer}, std::string::String>` is not implemented for `yew::virtual_dom::vcomp::VComp`
    |
    = help: the following implementations were found:
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, T>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, std::option::Option<T>>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::option::Option<std::string::String>>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::string::String>>
            and 3 others
    = note: required by `yew::virtual_dom::Transformer::transform`

error[E0308]: mismatched types
   --> $DIR/html-component-fail.rs:102:30
    |
102 |     html! { <Child int=1 ref=() /> };
    |                              ^^ expected struct `yew::html::NodeRef`, found `()`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VComp: yew::virtual_dom::Transformer<u32, i32>` is not satisfied
   --> $DIR/html-component-fail.rs:104:24
    |
104 |     html! { <Child int=0u32 /> };
    |                        ^^^^ the trait `yew::virtual_dom::Transformer<u32, i32>` is not implemented for `yew::virtual_dom::vcomp::VComp`
    |
    = help: the following implementations were found:
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, T>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a T, std::option::Option<T>>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::option::Option<std::string::String>>>
              <yew::virtual_dom::vcomp::VComp as yew::virtual_dom::Transformer<&'a str, std::string::String>>
            and 3 others
    = note: required by `yew::virtual_dom::Transformer::transform`

error[E0599]: no method named `string` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
   --> $DIR/html-component-fail.rs:105:20
    |
6   | #[derive(Clone, Properties, PartialEq)]
    |                 ---------- method `string` not found for this
...
105 |     html! { <Child string="abc" /> };
    |                    ^^^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `string`, perhaps you need to implement it:
            candidate #1: `proc_macro::bridge::server::Literal`

error[E0599]: no method named `children` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
   --> $DIR/html-component-fail.rs:109:14
    |
6   | #[derive(Clone, Properties, PartialEq)]
    |                 ---------- method `children` not found for this
...
109 |     html! { <Child>{ "Not allowed" }</Child> };
    |              ^^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`

error[E0599]: no method named `build` found for struct `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>` in the current scope
   --> $DIR/html-component-fail.rs:111:14
    |
32  | #[derive(Clone, Properties)]
    |                 ---------- method `build` not found for this
...
111 |     html! { <ChildContainer /> };
    |              ^^^^^^^^^^^^^^ method not found in `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `build`, perhaps you need to implement it:
            candidate #1: `proc_macro::bridge::server::TokenStreamBuilder`

error[E0599]: no method named `build` found for struct `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>` in the current scope
   --> $DIR/html-component-fail.rs:112:14
    |
32  | #[derive(Clone, Properties)]
    |                 ---------- method `build` not found for this
...
112 |     html! { <ChildContainer></ChildContainer> };
    |              ^^^^^^^^^^^^^^ method not found in `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `build`, perhaps you need to implement it:
            candidate #1: `proc_macro::bridge::server::TokenStreamBuilder`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VChild<Child>: std::convert::From<yew::virtual_dom::vtext::VText>` is not satisfied
   --> $DIR/html-component-fail.rs:113:31
    |
113 |     html! { <ChildContainer>{ "Not allowed" }</ChildContainer> };
    |                               ^^^^^^^^^^^^^ the trait `std::convert::From<yew::virtual_dom::vtext::VText>` is not implemented for `yew::virtual_dom::vcomp::VChild<Child>`
    |
    = note: required because of the requirements on the impl of `std::convert::Into<yew::virtual_dom::vcomp::VChild<Child>>` for `yew::virtual_dom::vtext::VText`
    = note: required by `std::convert::Into::into`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VChild<Child>: std::convert::From<yew::virtual_dom::vnode::VNode>` is not satisfied
   --> $DIR/html-component-fail.rs:114:29
    |
114 |     html! { <ChildContainer><></></ChildContainer> };
    |                             ^ the trait `std::convert::From<yew::virtual_dom::vnode::VNode>` is not implemented for `yew::virtual_dom::vcomp::VChild<Child>`
    |
    = note: required because of the requirements on the impl of `std::convert::Into<yew::virtual_dom::vcomp::VChild<Child>>` for `yew::virtual_dom::vnode::VNode`
    = note: required by `std::convert::Into::into`

error[E0277]: the trait bound `yew::virtual_dom::vcomp::VChild<Child>: std::convert::From<yew::virtual_dom::vnode::VNode>` is not satisfied
   --> $DIR/html-component-fail.rs:115:30
    |
115 |     html! { <ChildContainer><other /></ChildContainer> };
    |                              ^^^^^ the trait `std::convert::From<yew::virtual_dom::vnode::VNode>` is not implemented for `yew::virtual_dom::vcomp::VChild<Child>`
    |
    = note: required because of the requirements on the impl of `std::convert::Into<yew::virtual_dom::vcomp::VChild<Child>>` for `yew::virtual_dom::vnode::VNode`
    = note: required by `std::convert::Into::into`
